specifications/mempool/README.md

```diff
@@ -0,0 +1,280 @@
+```rust
+fn process_consensus_request(msg: ConsensusRequest)
+```
+Expected input:
+* msg: a message from consensus. There can be two types:
+    1. request for pulling a consensus block of transactions
+    2. notification of rejected committed transactions
+
+Required behavior:
+
+* if `msg` is q get-block request, return a response with consensus-ready transactions
+* if `msg` is notification of rejected txns, remove the transactions rejected from consensus from core mempool
+
+```rust
+fn process_state_sync_request(msg: CommitNotification)
+```
+
+Expected input:
+* msg: a message containing:
+    - transactions that were successfully committed
+    - timestamp of committed block
+
+Required behavior:
+
+* remove successfully committed transactions
+* clean out transactions with expiration time older than block timestamp
+* send `CommitResponse` to callback
+
+```rust
+fn process_client_transactions(txn: SignedTransaction, callback)
+```
+
+Expected input:
+
+* `txn` - transaction submitted by JSON RPC client proxied to mempool by JSON RPC service
+* `callback` - callback to JSON RPC service
+
+Required behavior:
+
+* validate and submit transaction to mempool
+* send back transaction submission result to callback
+
+```rust
+fn broadcast_single_peer(peer, peer_sync_data)
+```
+
+Expected input:
+
+* `peer` - peer to broadcast to
+* `peer_sync_data` - peer-specific information to determine which transactions to broadcast to this peer. This can contain information like the transaction broadcast history for this peer
+
+required:
+
+* crafts `MempoolSyncMsg::BroadcastTransactionsRequest` with broadcast-ready transactions for this peer
+* network-sends MempoolSyncMsg to peer
+
+```rust
+fn process_transaction_broadcast(msg: MempoolSyncMsg)
+```
+
+Expected input:
+
+* `MempoolSyncMsg::BroadcastTransactionsRequest`
+
+Required behavior:
+
+* validate and submit transactions to mempool
+* craft corresponding `MempoolSyncMsg::BroadcastTransactionsResponse` and send to peer (via `network::send_to(peer, msg)`)
+
+## Abstracted Modules
+
+### VM
+
+Mempool performs transaction signature verification and validations via the VM module. 
+
+```rust
+fn validate_transaction(&self, txn: SignedTransaction)
+```
+
+* [validates](#transaction-states-in-mempool) a transaction
+
+### Consensus
+
+Consensus interacts with mempool by (1) requesting a block of transactions
+ready for consensus, and (2) notifying mempool of rejected committed transactions
+- see [here](#functionality:-processing-consensus-requests) for the required 
+behavior for this interaction. Otherwise, implementing and coordinating this 
+interaction is an implementation detail.
+
+### State Sync
+
+Mempool receives notification from state sync about committed transactions -
+see [here](#functionality:-processing-state-sync-requests) for the required
+behavior for this interaction. Otherwise, implementing and coordinating 
+this interaction is an implementation detail.
```

Comment:
I think it might be more intuitive to describe the flow into and out of mempool.

* how does it arrive
  * directly, e.g., json-rpc
  * from downstream peers -> broadcast to other peers and upstream
  * from peers -> only sent upstream
* how does it leave
  * ttl
  * commit
  * discard
* how does it get consumed -- consensus

Reply:
sort of already cover this in the overview